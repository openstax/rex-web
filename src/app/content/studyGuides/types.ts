import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import {
  CountsPerSource,
  SummaryHighlights,
  SummaryHighlightsPagination,
} from '../highlights/types';

export interface StudyGuidesSummaryFilters {
  colors: HighlightColorEnum[];
  locationIds: string[];
}

export interface State {
  isEnabled: boolean;
  summary: {
    filters: StudyGuidesSummaryFilters,
    loading: boolean,
    open: boolean,
    pagination: SummaryHighlightsPagination,
    studyGuides: SummaryHighlights | null
    totalCountsPerPage: CountsPerSource | null;
  };
}
