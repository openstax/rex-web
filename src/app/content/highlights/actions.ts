import { Highlight, NewHighlight, UpdateHighlightRequest } from '@openstax/highlighter/dist/api';
import { createStandardAction } from 'typesafe-actions';
import { LocationFilters } from '../components/popUp/types';
import {
  CountsPerSource,
  HighlightData,
  SummaryFilters,
  SummaryFiltersUpdate,
  SummaryHighlights,
  SummaryHighlightsPagination,
} from './types';

export const focusHighlight = createStandardAction('Content/Highlights/focus')<string>();
export const clearFocusedHighlight = createStandardAction('Content/Highlights/clear')();
export const createHighlight = createStandardAction('Content/Highlights/create')<NewHighlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const requestDeleteHighlight = createStandardAction('Content/Highlights/requestDelete')<Highlight, {
  locationFilterId: string,
  pageId: string,
}>();
export const receiveDeleteHighlight = createStandardAction('Content/Highlights/receiveDelete')<Highlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const updateHighlight = createStandardAction('Content/Highlights/update')<UpdateHighlightRequest, {
  revertingAfterFailure?: boolean,
  preUpdateData: UpdateHighlightRequest,
  locationFilterId: string,
  pageId: string,
}>();
export const receiveHighlights = createStandardAction(
  'Content/Highlights/receive'
)<{highlights: HighlightData[], pageId: string}>();
export const setAnnotationChangesPending = createStandardAction('Content/Highlights/setAnnotationChangesPending')<
  boolean
>();

export const openMyHighlights = createStandardAction('Content/Highlights/Summary/open')<void>();
export const closeMyHighlights = createStandardAction('Content/Highlights/Summary/close')<void>();
export const initializeMyHighlightsSummary = createStandardAction('Content/Highlights/Summary/init')<void>();

export const printSummaryHighlights = createStandardAction('Content/Highlights/Summary/print')();
export const toggleSummaryHighlightsLoading = createStandardAction('Content/Highlights/Summary/loading')<boolean>();

export const loadMoreSummaryHighlights = createStandardAction('Content/Highlights/Summary/loadMore')();
export const setSummaryFilters = createStandardAction('Content/Highlights/Summary/setFilters')<
  Partial<SummaryFilters>
>();
export const updateSummaryFilters = createStandardAction('Content/Highlights/Summary/updateFilters')<
  Partial<SummaryFiltersUpdate>
>();
export const receiveSummaryHighlights = createStandardAction('Content/Highlights/Summary/receiveHighlights')<
  SummaryHighlights,
  {
    pagination: SummaryHighlightsPagination,
    filters?: SummaryFilters,
    isStillLoading?: boolean
  }
>();
export const receiveHighlightsTotalCounts = createStandardAction(
  'Content/receiveHighlightsTotalCounts'
)<CountsPerSource, LocationFilters>();
