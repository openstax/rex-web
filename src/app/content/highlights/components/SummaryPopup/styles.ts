import styled from 'styled-components/macro';

// tslint:disable-next-line:variable-name
export const HighlightEditButtons = styled.div`
  display: flex;
  align-items: center;
  padding: 0 8px;
  overflow: visible;

  button:first-child {
    margin-right: 8px;
  }
`;
