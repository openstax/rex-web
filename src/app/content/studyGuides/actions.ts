import { createStandardAction } from 'typesafe-actions';
import {
  CountsPerSource,
  SummaryFilters,
  SummaryFiltersUpdate,
  SummaryHighlights,
  SummaryHighlightsPagination
} from '../highlights/types';
import { StudyGuidesSummaryFilters } from './types';

export const receiveSummaryStudyGuides = createStandardAction(
  'Content/StudyGuides/Summary/receive'
)<
SummaryHighlights,
  {
    pagination: SummaryHighlightsPagination,
    filters?: StudyGuidesSummaryFilters,
    isStillLoading?: boolean
  }
>();

export const openStudyGuides = createStandardAction('Content/StudyGuides/Summary/open')<void>();
export const closeStudyGuides = createStandardAction('Content/StudyGuides/Summary/close')<void>();
export const loadMoreStudyGuides = createStandardAction('Content/StudyGuides/loadMore')();
export const printStudyGuides = createStandardAction('Content/StudyGuides/print')<void>();
export const setDefaultSummaryFilters = createStandardAction('Content/StudyGuides/Summary/setDefaultFilters')<
  SummaryFilters
>();
export const setSummaryFilters = createStandardAction('Content/StudyGuides/Summary/setFilters')<
  Partial<SummaryFilters>
>();
export const updateSummaryFilters = createStandardAction('Content/StudyGuides/Summary/updateFilters')<
  SummaryFiltersUpdate
>();
export const toggleStudyGuidesSummaryLoading = createStandardAction('Content/StudyGuides/Summary/loading')<boolean>();
export const receiveStudyGuidesTotalCounts = createStandardAction(
  'Content/StudyGuides/receiveTotalCounts'
)<CountsPerSource>();
