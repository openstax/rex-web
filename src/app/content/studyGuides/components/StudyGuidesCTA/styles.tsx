import styled, { css } from 'styled-components/macro';
import Button from '../../../../components/Button';
import theme from '../../../../theme';
import { disablePrint } from '../../../components/utils/disablePrint';

// tslint:disable-next-line: variable-name
export const StudyGuidesCTAWrapper = styled.div`
  display: flex;
  height: 20rem;
  padding: 0 8rem;
  justify-content: space-between;
  align-items: center;
  color: ${theme.color.text.default};
  box-shadow: inset 0 0 10px 0 rgba(0, 0, 0, 0.4);
  overflow: hidden;

  strong {
    white-space: nowrap;
    color: ${theme.color.secondary.deepGreen.base};
  }

  ${theme.breakpoints.mobile(css`
    height: 17.6rem;
    padding: 0 1.6rem;
    justify-content: center;
  `)}

  ${disablePrint}
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTAInnerWrapper = styled.div`
  display: contents;
  ${theme.breakpoints.mobile(css`
    display: flex;
    max-width: 34rem;
    height: 100%;
    overflow: visible;
  `)}
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTAContent = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTATitle = styled.h2`
  overflow: hidden;
  font-size: 2.4rem;
  margin: 0 0 1rem 0;
  ${theme.breakpoints.mobile(css`
    font-size: 1.8rem;
    line-height: 2.6rem;
    margin-top: 1rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTAButtons = styled.div`
  overflow: hidden;
  display: flex;
  align-items: center;
  flex-wrap: wrap;
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTAButton = styled(Button)`
  width: 24rem;
  height: 4.8rem;
  padding: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 1.8rem;
  font-weight: 600;
  color: ${theme.color.text.white};
  background-color: ${theme.color.secondary.deepGreen.base};
  border: none;
  margin-right: 1.6rem;

  &:hover {
    background-color: ${theme.color.secondary.deepGreen.base};
  }

  ${theme.breakpoints.mobile(css`
    font-size: 1.4rem;
    width: 11rem;
    height: 3rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTAButtonsSecondary = styled.div`
  padding: 1.4rem 0;
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTASeparator = styled.span`
  font-size: 1.6rem;
  text-transform: uppercase;
  margin-right: 1.6rem;
  ${theme.breakpoints.mobile(css`
    font-size: 1.4rem;
    margin-right: 1.2rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTALink = styled.a`
  font-size: 1.6rem;
  color: #027eb5;
  ${theme.breakpoints.mobile(css`
    font-size: 1.4rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTAInfoWrapper = styled.div`
  flex: 1;
  display: flex;
  justify-content: center;
  height: 100%;
  text-align: center;
  overflow: visible;
  position: relative;
  ${theme.breakpoints.mobile(css`
    justify-content: flex-start;
  `)}
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTAInfo = styled.div`
  font-size: 1.8rem;
  line-height: 2.5rem;
  width: 24rem;
  margin-top: 2.5rem;
  z-index: 1;

  strong {
    display: block;
  }

  ${theme.breakpoints.mobile(css`
    width: 10rem;
    font-size: 1.2rem;
    line-height: 1.7rem;
    margin-left: 2rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTAArrowDesktop = styled.img`
  position: absolute;
  ${theme.breakpoints.mobile(css`
    display: none;
  `)}
`;

// tslint:disable-next-line: variable-name
export const StudyGuidesCTAArrowMobile = styled.img`
  position: absolute;
  left: -4rem;
  display: none;
  ${theme.breakpoints.mobile(css`
    display: block;
  `)}
`;
