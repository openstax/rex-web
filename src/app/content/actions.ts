import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as practiceQuestionsActions from './practiceQuestions/actions';
import * as searchActions from './search/actions';
import * as studyGuidesActions from './studyGuides/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;
export const studyGuides = studyGuidesActions;
export const practiceQuestions = practiceQuestionsActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const receiveBuyPrintConfig = createStandardAction('Content/receiveBuyPrintConfig')<State['buyPrint']>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();
export const receivePageNotFoundId = createStandardAction('Content/receivePageNotFoundId')<string>();

export const openNudgeStudyTools = createStandardAction('Content/openNudgeStudyTools')();
export const closeNudgeStudyTools = createStandardAction('Content/closeNudgeStudyTools')();
