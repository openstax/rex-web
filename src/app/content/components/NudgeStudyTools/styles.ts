import styled, { css } from 'styled-components/macro';
import { Times } from 'styled-icons/fa-solid/Times/Times';
import { PlainButton } from '../../../components/Button';
import htmlMessage from '../../../components/htmlMessage';
import theme from '../../../theme';
import { remsToPx } from '../../../utils';
import { arrowDesktopHeight, closeButtonMobileMargin, closeButtonSize, contentWidth } from './constants';

// tslint:disable-next-line: variable-name
export const NudgeWrapper = styled.div`
  display: contents;
`;

// tslint:disable-next-line: variable-name
export const NudgeContentWrapper = styled.div`
  position: fixed;
  width: ${contentWidth};
  z-index: ${theme.zIndex.nudgeOverlay + 1};
  outline: none;
  ${(props: { top: number, left: number }) => `
    top: ${props.top}px;
    left: ${props.left}px;
  `}
  ${theme.breakpoints.mobileMedium(css`
    left: auto;
    width: 100%;
    padding: 0 1.6rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeContent = styled.div`
  position: relative;
`;

// tslint:disable-next-line: variable-name
const NudgeHeadingStyles = styled.h2`
  font-size: 3.6rem;
  line-height: 1.1;
  letter-spacing: -1.4px;
  margin: 0 0 1.6rem 0;
  color: ${theme.color.text.white};
  overflow: hidden;

  strong {
    color: ${theme.color.primary.yellow.base};
    font-weight: 600;
    white-space: nowrap;
  }

  ${theme.breakpoints.mobileMedium(css`
    font-size: 2.4rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeHeading = htmlMessage('i18n:nudge:study-tools:heading', NudgeHeadingStyles);

// tslint:disable-next-line: variable-name
export const NudgeTextStyles = styled.div`
  font-size: 2.4rem;
  font-weight: 400;
  line-height: 1.1;
  letter-spacing: -1px;
  color: ${theme.color.text.white};
  max-width: 690px;
  overflow: hidden;
  ${theme.breakpoints.mobileMedium(css`
    max-width: 100%;
    font-size: 1.6rem;
    line-height: 1.4;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeArrow = styled.img`
  position: fixed;
  z-index: ${theme.zIndex.nudgeOverlay + 1};
  display: block;
  height: ${arrowDesktopHeight}rem;
  ${(props: { top: number, left: number }) => `
    top: ${props.top}px;
    left: ${props.left}px;
  `}
`;

// tslint:disable-next-line: variable-name
export const NudgeCloseIcon = styled(Times)`
  width: 1.1rem;
  color: ${theme.color.text.white};
`;

// tslint:disable-next-line: variable-name
export const NudgeCloseButton = styled(PlainButton)`
  position: fixed;
  z-index: ${theme.zIndex.nudgeOverlay + 1};
  ${(props: { top: number, left: number }) => `
    top: ${props.top}px;
    left: ${props.left}px;
  `}
  width: ${remsToPx(closeButtonSize)}px;
  height: ${remsToPx(closeButtonSize)}px;
  padding: 1rem;
  border-radius: 50%;
  border: 1px solid ${theme.color.white};
  display: flex;
  align-items: center;
  justify-content: center;

  ${theme.breakpoints.mobile(css`
    left: auto;
    top: ${remsToPx(closeButtonMobileMargin)}px;
    right: ${remsToPx(closeButtonMobileMargin)}px;
  `)}
`;

interface NudgeSpotlightPlacement {
  top: number;
  left: number;
  height: number;
  width: number;
}

// tslint:disable-next-line: variable-name
export const NudgeBackground = styled.div`
  position: fixed;
  top: 0;
  width: 100%;
  height: 100%;
  z-index: ${theme.zIndex.nudgeOverlay};
  pointer-events: none;

  display: grid;
  ${(props: NudgeSpotlightPlacement) => `
    grid-template: ${props.top}px ${props.height}px 1fr / ${props.left}px ${props.width}px 1fr;
    grid-template-areas:
      'top       top       top'
      'left   spotlight   right'
      'bottom   bottom   bottom';
  `}
`;

interface ClickBlockerProps {
  area: 'top' | 'right' | 'bottom' | 'left';
}

// tslint:disable-next-line: variable-name
export const ClickBlocker = styled.div`
  width: 100%;
  height: 100%;
  pointer-events: all;
  background-color: ${theme.color.black};
  opacity: 0.9;

  grid-area: ${({area}: ClickBlockerProps) => area};
`;

// tslint:disable-next-line: variable-name
export const NudgeElementTarget = styled.div`
  display: contents;
`;
