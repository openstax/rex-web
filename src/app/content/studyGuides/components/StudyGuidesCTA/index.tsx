import React from 'react';
import { FormattedMessage } from 'react-intl';
import { useSelector } from 'react-redux';
import {
  loggedOut,
  loginLink as loginLinkSelector,
  signupLink as signupLinkSelector,
} from '../../../../auth/selectors';
import htmlMessage from '../../../../components/htmlMessage';
import arrowDesktop from './assets/arrow-desktop.svg';
import arrowMobile from './assets/arrow-mobile.svg';
import * as Styled from './styles';

// tslint:disable-next-line:variable-name
const StudyGuidesCTATitle = htmlMessage('i18n:studyguides:cta:title', Styled.StudyGuidesCTATitle);

// tslint:disable-next-line:variable-name
const StudyGuidesCTAInfo = htmlMessage('i18n:studyguides:cta:info', Styled.StudyGuidesCTAInfo);

// tslint:disable-next-line: variable-name
const StudyGuidesCTA = () => {
  const isNotLoggedIn = useSelector(loggedOut);
  const loginLink = useSelector(loginLinkSelector);
  const signupLink = useSelector(signupLinkSelector);

  if (!isNotLoggedIn) { return null; }

  return <Styled.StudyGuidesCTAWrapper>
    <Styled.StudyGuidesCTAInnerWrapper>
      <Styled.StudyGuidesCTAContent>
        <StudyGuidesCTATitle />
        <Styled.StudyGuidesCTAButtons>
          <FormattedMessage id='i18n:studyguides:cta:button'>
            {(msg) => <Styled.StudyGuidesCTAButton
              // eslint-disable-next-line jsx-a11y/anchor-has-content
              component={<a href={signupLink} data-analytics-label='signup' />}
            >{msg}</Styled.StudyGuidesCTAButton>}
          </FormattedMessage>
          <Styled.StudyGuidesCTAButtonsSecondary>
            <Styled.StudyGuidesCTASeparator>
              <FormattedMessage id='i18n:studyguides:cta:separator'>
                {(msg) => msg}
              </FormattedMessage>
            </Styled.StudyGuidesCTASeparator>
            <FormattedMessage id='i18n:studyguides:cta:login'>
              {(msg) => <Styled.StudyGuidesCTALink
                data-analytics-label='login'
                href={loginLink}
              >{msg}</Styled.StudyGuidesCTALink>}
            </FormattedMessage>
          </Styled.StudyGuidesCTAButtonsSecondary>
        </Styled.StudyGuidesCTAButtons>
      </Styled.StudyGuidesCTAContent>
      <Styled.StudyGuidesCTAInfoWrapper>
        <StudyGuidesCTAInfo />
        <Styled.StudyGuidesCTAArrowDesktop src={arrowDesktop} alt='' />
        <Styled.StudyGuidesCTAArrowMobile src={arrowMobile} alt='' />
      </Styled.StudyGuidesCTAInfoWrapper>
    </Styled.StudyGuidesCTAInnerWrapper>
  </Styled.StudyGuidesCTAWrapper>;
};

export default StudyGuidesCTA;
