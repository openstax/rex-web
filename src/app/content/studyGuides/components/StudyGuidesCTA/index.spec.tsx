import React from 'react';
import renderer from 'react-test-renderer';
import StudyGuidesCTA from '.';
import createTestStore from '../../../../../test/createTestStore';
import TestContainer from '../../../../../test/TestContainer';
import { receiveLoggedOut, receiveUser } from '../../../../auth/actions';
import { Store } from '../../../../types';
import * as Styled from './styles';

describe('StudyGuidesCTA', () => {
  let store: Store;

  beforeEach(() => {
    store = createTestStore();
  });

  it('does not render if user is logged in', () => {
    store.dispatch(receiveUser({} as any));

    const component = renderer.create(<TestContainer store={store}>
      <StudyGuidesCTA />
    </TestContainer>);

    expect(() => component.root.findByType(Styled.StudyGuidesCTAWrapper)).toThrow();
  });

  it('does render if user is not logged in', () => {
    store.dispatch(receiveLoggedOut());

    const component = renderer.create(<TestContainer store={store}>
      <StudyGuidesCTA />
    </TestContainer>);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
