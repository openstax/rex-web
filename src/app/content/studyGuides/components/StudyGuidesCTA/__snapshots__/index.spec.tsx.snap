// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudyGuidesCTA does render if user is not logged in 1`] = `
<div
  className="styles__StudyGuidesCTAWrapper-dthmzf-0 ldAdwo"
>
  <div
    className="styles__StudyGuidesCTAInnerWrapper-dthmzf-1 ekANVM"
  >
    <div
      className="styles__StudyGuidesCTAContent-dthmzf-2 hZLeDx"
    >
      <h2
        className="styles__StudyGuidesCTATitle-dthmzf-3 igvIfZ"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Expert-created study guides. <strong>100% free!</strong>",
          }
        }
      />
      <div
        className="styles__StudyGuidesCTAButtons-dthmzf-4 bESOcY"
      >
        <a
          className="Button-ayg7nk-0 styles__StudyGuidesCTAButton-dthmzf-5 kfWnpM"
          data-analytics-label="signup"
          href="/accounts/signup?r=/"
        >
          Sign Up
        </a>
        <div
          className="styles__StudyGuidesCTAButtonsSecondary-dthmzf-6 dSPDZh"
        >
          <span
            className="styles__StudyGuidesCTASeparator-dthmzf-7 bXhklW"
          >
            or
          </span>
          <a
            className="styles__StudyGuidesCTALink-dthmzf-8 hLoOdb"
            data-analytics-label="login"
            href="/accounts/login?r=/"
          >
            Log in
          </a>
        </div>
      </div>
    </div>
    <div
      className="styles__StudyGuidesCTAInfoWrapper-dthmzf-9 iGpeCA"
    >
      <div
        className="styles__StudyGuidesCTAInfo-dthmzf-10 dKsjzx"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Like this guide? <br>We've got study guides for every chapter. <strong>Free forever.</strong>",
          }
        }
      />
      <img
        alt=""
        className="styles__StudyGuidesCTAArrowDesktop-dthmzf-11 cKiziA"
        src="arrow-desktop.svg"
      />
      <img
        alt=""
        className="styles__StudyGuidesCTAArrowMobile-dthmzf-12 itsJER"
        src="arrow-mobile.svg"
      />
    </div>
  </div>
</div>
`;
